File: src/lib/converter/character/wowhead-exporter/character-direct.ts
/**
 * Converter-side direct character export (WOW_PIPELINE=direct).
 *
 * Replaces the /rest/exportCharacter RPC: fetches DB2-derived metadata from
 * /rest/charMeta, builds the geoset mask and bakes customization materials
 * locally (CharMaterialRenderer over the raw-file layer), then runs the direct
 * M2 -> MDL conversion. Mirrors exportCharacterModelHeadless byte-for-byte.
 */
import { createHash } from 'crypto';
import path from 'path';

import { ExportContext } from '@/lib/converter/character/utils';
import { DirectDataTexture } from '@/lib/converter/wow-model/direct/m2/textures';
import { profileScope } from '@/lib/export-profile';
import { ensureConverterCasc } from '@/lib/wow/archive/client/remote-casc';
import { CharMaterialRenderer } from '@/lib/wow/character/char-material-renderer';
import type { GeosetMaskEntry } from '@/lib/wow/export/m2/m2-exporter';
import { replaceExtension } from '@/lib/wow/export/writers/export-helper';
import { Skin } from '@/lib/wow/formats/m2/skin';
import {
  CharacterMetaResponse, ExportCharacterParams, wowExportClient,
} from '@/lib/wowexport-client/wowexport-client';

import { Model } from '../../common/models';

/** Default-enabled geosets, mirroring the UI defaults (see M2Renderer). */
const DEFAULT_GEOSETS = new Set([0, 101, 201, 301, 401, 501, 601, 702, 801, 901, 1001, 1101, 1201, 1301, 1400, 1501, 1600, 1700, 1801, 1901, 2001, 2101, 2201, 2301, 2400, 2500, 2601, 2700, 2801, 2900, 3000, 3100, 3202, 3301, 3401, 3500, 3600, 3700, 3801, 3900, 4001, 4101, 4201, 4301, 4401, 4501, 4601, 4701, 4801, 4901, 5001, 5101]);

/** Port of the geoset mask construction in exportCharacterModelHeadless. */
function buildCharacterGeosetMask(
  skin: Skin,
  meta: CharacterMetaResponse,
  body: ExportCharacterParams,
): GeosetMaskEntry[] {
  const subMeshes = skin.subMeshes || [];
  const geosetGroup = (id: number): number => Math.floor(id / 100) * 100;

  const geosetMask = subMeshes.map((subMesh) => ({
    id: subMesh.submeshID,
    checked: DEFAULT_GEOSETS.has(subMesh.submeshID),
  }));

  const turnOnGeoset = (subMeshId: number, turnOffOthers = true): void => {
    const group = geosetGroup(subMeshId);
    const matchingGeosets = geosetMask.filter((geoset) => geoset.id === subMeshId);
    if (matchingGeosets.length > 0) {
      matchingGeosets.forEach((geoset) => { geoset.checked = true; });
      if (group === 0 || !turnOffOthers) return; // base geometry cannot be overridden
      geosetMask.forEach((geoset) => {
        if (group === geosetGroup(geoset.id) && geoset.id !== subMeshId) geoset.checked = false;
      });
    }
  };

  const hideGeosetIds = body.hideGeosetIds ?? [];

  // Turn on customization geosets
  for (const [, choiceID] of Object.entries(body.customizations || {})) {
    const geosets = meta.choices[Number(choiceID)]?.geosets ?? [];
    for (const geosetId of geosets) {
      if (!hideGeosetIds.includes(geosetId)) turnOnGeoset(geosetId, true);
    }
  }

  if (Array.isArray(body.geosetIds) && body.geosetIds.length > 0) {
    for (const geosetId of body.geosetIds) turnOnGeoset(geosetId, true);
  }

  if (Array.isArray(hideGeosetIds) && hideGeosetIds.length > 0) {
    const idsToHide = new Set(hideGeosetIds);
    for (const geoset of geosetMask) {
      if (idsToHide.has(geoset.id)) geoset.checked = false;
    }
  }

  return geosetMask;
}

/**
 * Baked materials are pure functions of (race, gender, model override,
 * customizations); cache the composited PNGs so re-exports of the same
 * character (e.g. mount + rider, warmup passes) skip the bake entirely.
 * Consumers only read the cached buffers (resolveTextures registers/writes
 * them), so sharing across requests is safe.
 */
const BAKE_CACHE_MAX = 8;
const bakeCache = new Map<string, Map<number, DirectDataTexture>>();

export function clearCharacterBakeCache(): void {
  bakeCache.clear();
}

function bakeCacheKey(body: ExportCharacterParams): string {
  return JSON.stringify({
    race: body.race,
    gender: body.gender,
    fileDataIdOverride: body.fileDataIdOverride ?? null,
    customizations: body.customizations ?? {},
  });
}

/** Port of the material bake loop in exportCharacterModelHeadless. */
async function bakeCharacterMaterials(
  meta: CharacterMetaResponse,
  body: ExportCharacterParams,
): Promise<Map<number, DirectDataTexture>> {
  CharMaterialRenderer.init();
  const chrMaterials = new Map<number, CharMaterialRenderer>();

  for (const [, choiceID] of Object.entries(body.customizations || {})) {
    const materials = meta.choices[Number(choiceID)]?.materials ?? [];
    for (const mat of materials) {
      let chrMaterial = chrMaterials.get(mat.material.TextureType);
      if (!chrMaterial) {
        chrMaterial = new CharMaterialRenderer(mat.material.TextureType, mat.material.Width, mat.material.Height, true);
        chrMaterials.set(mat.material.TextureType, chrMaterial);
        chrMaterial.init();
      }
      await chrMaterial.setTextureTarget(
        mat.custMaterial,
        mat.section!,
        mat.material,
        mat.textureLayer,
        true,
        mat.filename ?? undefined,
      );
    }
  }

  const dataTextures = new Map<number, DirectDataTexture>();
  for (const [textureType, chrMaterial] of chrMaterials) {
    let originalFilename: string | null = null;
    if (chrMaterial.textureTargets && chrMaterial.textureTargets.length > 0) {
      const target = chrMaterial.textureTargets.find((t) => t.filename);
      if (target && target.filename) originalFilename = target.filename;
    }
    dataTextures.set(Number(textureType), {
      filename: originalFilename,
      source: { kind: 'png', png: chrMaterial.getPNG() },
    });
    chrMaterial.dispose();
  }

  return dataTextures;
}

/** Mirror of the export path naming in exportCharacterModelHeadless + rest-server suffix. */
function resolveCharacterExportPath(exportRoot: string, fileName: string, exportSuffix: string): string {
  let exportPath = replaceExtension(path.normalize(path.join(exportRoot, fileName.replace(/\s/g, ''))), '.obj');
  if (exportSuffix) {
    const dir = path.dirname(exportPath);
    const base = path.basename(exportPath, path.extname(exportPath));
    const ext = path.extname(exportPath);
    exportPath = path.join(dir, `${base}_${exportSuffix}${ext}`);
  }
  return exportPath;
}

/**
 * Direct replacement for wowExportClient.exportCharacter + OBJ parse.
 * `body` must be the exact object the legacy RPC would have posted — the
 * export suffix (part of the model name) is derived from its JSON form.
 */
export async function exportCharacterDirectAsModel(ctx: ExportContext, body: ExportCharacterParams): Promise<Model> {
  ensureConverterCasc();

  const meta = await wowExportClient.getCharMeta({
    race: body.race,
    gender: body.gender,
    fileDataIdOverride: body.fileDataIdOverride,
    customizations: body.customizations,
  });

  const cacheKey = bakeCacheKey(body);
  let dataTextures = bakeCache.get(cacheKey);
  if (dataTextures) {
    // Refresh LRU recency.
    bakeCache.delete(cacheKey);
    bakeCache.set(cacheKey, dataTextures);
  } else {
    dataTextures = await profileScope('bakeMaterials', () => bakeCharacterMaterials(meta, body));
    bakeCache.set(cacheKey, dataTextures);
    if (bakeCache.size > BAKE_CACHE_MAX) bakeCache.delete(bakeCache.keys().next().value!);
  }

  // Same suffix the server derives from the RPC body (md5 of its JSON).
  const suffix = createHash('md5').update(JSON.stringify(body || {})).digest('hex').slice(0, 8);
  const exportPath = resolveCharacterExportPath(ctx.config.wowExportAssetDir, meta.fileName, suffix);

  return ctx.assetManager.parseDirect({
    fileDataID: meta.fileDataID,
    variantTextures: [],
    geosetMaskBuilder: (skin) => buildCharacterGeosetMask(skin, meta, body),
    dataTextures,
    excludeAnimIds: body.excludeAnimationIds ?? [],
    exportPathOverride: exportPath,
  });
}

File: golang/internal/converter/character/character_direct.go
package character

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"path/filepath"
	"sort"
	"strings"
	"sync"

	"github.com/pqhuy98/wow-converter/internal/converter/common"
	"github.com/pqhuy98/wow-converter/internal/converter/runtimecache"
	"github.com/pqhuy98/wow-converter/internal/converter/texturesource"
	directm2 "github.com/pqhuy98/wow-converter/internal/converter/wowmodel/direct/m2"
	"github.com/pqhuy98/wow-converter/internal/wow/casc"
	wowchar "github.com/pqhuy98/wow-converter/internal/wow/character"
	m2export "github.com/pqhuy98/wow-converter/internal/wow/export/m2"
	"github.com/pqhuy98/wow-converter/internal/wow/export/writers"
	"github.com/pqhuy98/wow-converter/internal/wow/formats/m2"
)

var defaultGeosets = map[int]struct{}{
	0: {}, 101: {}, 201: {}, 301: {}, 401: {}, 501: {}, 601: {}, 702: {}, 801: {}, 901: {},
	1001: {}, 1101: {}, 1201: {}, 1301: {}, 1400: {}, 1501: {}, 1600: {}, 1700: {}, 1801: {},
	1901: {}, 2001: {}, 2101: {}, 2201: {}, 2301: {}, 2400: {}, 2500: {}, 2601: {}, 2700: {},
	2801: {}, 2900: {}, 3000: {}, 3100: {}, 3202: {}, 3301: {}, 3401: {}, 3500: {}, 3600: {},
	3700: {}, 3801: {}, 3900: {}, 4001: {}, 4101: {}, 4201: {}, 4301: {}, 4401: {}, 4501: {},
	4601: {}, 4701: {}, 4801: {}, 4901: {}, 5001: {}, 5101: {},
}

// ExportCharacterParams describes the direct character export request body.
type ExportCharacterParams struct {
	Race                int            `json:"race"`
	Gender              int            `json:"gender"`
	FileDataIDOverride  *int           `json:"fileDataIdOverride,omitempty"`
	Customizations       map[string]int `json:"customizations"`
	CustomizationOrder   []int          `json:"-"`
	ExcludeAnimationIDs []int          `json:"excludeAnimationIds,omitempty"`
	GeosetIDs           []int          `json:"geosetIds,omitempty"`
	HideGeosetIDs       []int          `json:"hideGeosetIds,omitempty"`
}

type parsedChoiceMeta struct {
	Geosets   []int
	Materials []wowchar.CharMetaChoiceMaterial
}

const bakeCacheMax = 8

var (
	bakeCache   = map[string]map[int]directm2.DirectDataTexture{}
	bakeCacheMu sync.Mutex
)

// ClearCharacterBakeCache clears the material bake LRU cache.
func ClearCharacterBakeCache() {
	bakeCacheMu.Lock()
	bakeCache = map[string]map[int]directm2.DirectDataTexture{}
	bakeCacheMu.Unlock()
}

// BakeCacheStats returns character material bake cache sizes.
func BakeCacheStats() (entries int, textureSlots int) {
	bakeCacheMu.Lock()
	defer bakeCacheMu.Unlock()
	entries = len(bakeCache)
	for _, textures := range bakeCache {
		textureSlots += len(textures)
	}
	return entries, textureSlots
}

func init() {
	runtimecache.RegisterConverterClearHook(ClearCharacterBakeCache)
	runtimecache.RegisterMemoryStatHook(func() map[string]any {
		entries, slots := BakeCacheStats()
		return map[string]any{
			"characterBakeEntries":      entries,
			"characterBakeTextureSlots": slots,
		}
	})
	runtimecache.RegisterMemoryStatHook(func() map[string]any {
		entries, pngBytes := texturesource.CacheStats()
		return map[string]any{
			"textureRegistryEntries":  entries,
			"textureRegistryPngBytes": pngBytes,
		}
	})
}

func buildCharacterGeosetMask(skin *m2.Skin, choices map[int]parsedChoiceMeta, body ExportCharacterParams) []m2export.GeosetMaskEntry {
	subMeshes := skin.SubMeshes
	geosetGroup := func(id int) int { return (id / 100) * 100 }
	mask := make([]m2export.GeosetMaskEntry, len(subMeshes))
	for i, sub := range subMeshes {
		id := int(sub.SubmeshID)
		_, checked := defaultGeosets[id]
		mask[i] = m2export.GeosetMaskEntry{ID: id, Checked: checked}
	}
	turnOn := func(subMeshID int, turnOffOthers bool) {
		group := geosetGroup(subMeshID)
		found := false
		for i := range mask {
			if mask[i].ID == subMeshID {
				mask[i].Checked = true
				found = true
			}
		}
		if !found || group == 0 || !turnOffOthers {
			return
		}
		for i := range mask {
			if geosetGroup(mask[i].ID) == group && mask[i].ID != subMeshID {
				mask[i].Checked = false
			}
		}
	}
	hide := body.HideGeosetIDs
	for _, choiceID := range customizationChoiceOrder(body) {
		if choice, ok := choices[choiceID]; ok {
			for _, geosetID := range choice.Geosets {
				if !containsInt(hide, geosetID) {
					turnOn(geosetID, true)
				}
			}
		}
	}
	for _, geosetID := range body.GeosetIDs {
		turnOn(geosetID, true)
	}
	hideSet := map[int]struct{}{}
	for _, id := range hide {
		hideSet[id] = struct{}{}
	}
	for i := range mask {
		if _, ok := hideSet[mask[i].ID]; ok {
			mask[i].Checked = false
		}
	}
	return mask
}

// ExportCharacterDirectAsModel runs the direct character M2 -> MDL pipeline.
func ExportCharacterDirectAsModel(ctx *ExportContext, body ExportCharacterParams) (*common.Model, error) {
	meta, choices, err := resolveCharMeta(ctx, body)
	if err != nil {
		return nil, err
	}
	cacheKey := bakeCacheKey(body)
	bakeCacheMu.Lock()
	dataTextures, ok := bakeCache[cacheKey]
	if ok {
		delete(bakeCache, cacheKey)
		bakeCache[cacheKey] = dataTextures
	}
	bakeCacheMu.Unlock()
	if !ok {
		dataTextures, err = bakeCharacterMaterials(ctx, choices, body)
		if err != nil {
			return nil, err
		}
		bakeCacheMu.Lock()
		bakeCache[cacheKey] = dataTextures
		for len(bakeCache) > bakeCacheMax {
			for k := range bakeCache {
				delete(bakeCache, k)
				break
			}
		}
		bakeCacheMu.Unlock()
	}

	bodyJSON, _ := json.Marshal(body)
	sum := md5.Sum(bodyJSON)
	suffix := hex.EncodeToString(sum[:])[:8]
	exportPath := resolveCharacterExportPath(ctx.Config.ExportAssetDir, meta.FileName, suffix)

	exclude := map[int]struct{}{}
	for _, id := range body.ExcludeAnimationIDs {
		exclude[id] = struct{}{}
	}

	return ctx.AssetManager.ParseDirectOptions(context.Background(), common.DirectParseOptions{
		FileDataID: meta.FileDataID,
		GeosetMaskBuilder: func(skin *m2.Skin) []m2export.GeosetMaskEntry {
			return buildCharacterGeosetMask(skin, choices, body)
		},
		DataTextures:       dataTextures,
		ExcludeAnimIDs:     exclude,
		ExportPathOverride: exportPath,
	})
}

func resolveCharMeta(ctx *ExportContext, body ExportCharacterParams) (wowchar.MetaResult, map[int]parsedChoiceMeta, error) {
	if ctx.WowClient == nil {
		return wowchar.MetaResult{}, nil, fmt.Errorf("wow client required for character export")
	}
	params := casc.CharacterMetaParams{
		Race: body.Race, Gender: body.Gender,
		FileDataIDOverride: body.FileDataIDOverride,
		Customizations:     body.Customizations,
	}
	resp, err := ctx.WowClient.GetCharMeta(context.Background(), params)
	if err != nil {
		return wowchar.MetaResult{}, nil, err
	}
	meta := wowchar.MetaResult{
		FileDataID: resp.FileDataID, FileName: resp.FileName, TextureLayoutID: resp.TextureLayoutID,
	}
	choices := map[int]parsedChoiceMeta{}
	for key, raw := range resp.Choices {
		var choiceID int
		fmt.Sscanf(key, "%d", &choiceID)
		data, _ := json.Marshal(raw)
		var parsed wowchar.ChoiceMeta
		if err := json.Unmarshal(data, &parsed); err != nil {
			continue
		}
		choices[choiceID] = parsedChoiceMeta{Geosets: parsed.Geosets, Materials: parsed.Materials}
		meta.Choices = map[int]wowchar.ChoiceMeta{choiceID: parsed}
	}
	return meta, choices, nil
}

func bakeCacheKey(body ExportCharacterParams) string {
	b, _ := json.Marshal(map[string]any{
		"race": body.Race, "gender": body.Gender,
		"fileDataIdOverride": body.FileDataIDOverride,
		"customizations":     body.Customizations,
	})
	return string(b)
}

func bakeCharacterMaterials(ctx *ExportContext, choices map[int]parsedChoiceMeta, body ExportCharacterParams) (map[int]directm2.DirectDataTexture, error) {
	loader := wowchar.TextureLoader(func(c context.Context, fileDataID int) ([]byte, error) {
		return ctx.WowClient.DownloadCascFile(c, fileDataID)
	})
	chrMaterials := map[int]*wowchar.MaterialRenderer{}
	chrMaterialOrder := []int{}
	for _, choiceID := range customizationChoiceOrder(body) {
		choice, ok := choices[choiceID]
		if !ok {
			continue
		}
		for _, mat := range choice.Materials {
			if mat.Section == nil {
				continue
			}
			texType := mat.Material.TextureType
			renderer := chrMaterials[texType]
			if renderer == nil {
				renderer = wowchar.NewMaterialRenderer(texType, mat.Material.Width, mat.Material.Height, loader)
				chrMaterials[texType] = renderer
				chrMaterialOrder = append(chrMaterialOrder, texType)
				renderer.Init()
			}
			filename := ""
			if mat.Filename != nil {
				filename = *mat.Filename
			}
			if err := renderer.SetTextureTarget(context.Background(), mat.CustMaterial, *mat.Section, mat.Material, mat.TextureLayer, true, filename); err != nil {
				return nil, err
			}
		}
	}
	out := map[int]directm2.DirectDataTexture{}
	for order, texType := range chrMaterialOrder {
		renderer := chrMaterials[texType]
		var filename *string
		pngBytes, err := renderer.GetPNG()
		if err != nil {
			return nil, err
		}
		if fn := renderer.FirstFilename(); fn != "" {
			filename = &fn
		}
		out[texType] = directm2.DirectDataTexture{
			Filename: filename,
			Source:   texturesource.Source{Kind: texturesource.KindPNG, PNG: pngBytes},
			Order:    order,
		}
		renderer.Dispose()
	}
	return out, nil
}

func customizationChoiceOrder(body ExportCharacterParams) []int {
	if len(body.CustomizationOrder) > 0 {
		return body.CustomizationOrder
	}
	return sortedCustomizationChoices(body.Customizations)
}

func sortedCustomizationChoices(customizations map[string]int) []int {
	keys := make([]string, 0, len(customizations))
	for key := range customizations {
		keys = append(keys, key)
	}
	sort.Slice(keys, func(i, j int) bool {
		ki, kj := 0, 0
		fmt.Sscanf(keys[i], "%d", &ki)
		fmt.Sscanf(keys[j], "%d", &kj)
		return ki < kj
	})
	out := make([]int, 0, len(keys))
	for _, key := range keys {
		out = append(out, customizations[key])
	}
	return out
}

func resolveCharacterExportPath(exportRoot, fileName, suffix string) string {
	exportPath := writers.ReplaceExtension(filepath.Join(exportRoot, strings.ReplaceAll(fileName, " ", "")), ".m2")
	if suffix == "" {
		return exportPath
	}
	dir := filepath.Dir(exportPath)
	base := strings.TrimSuffix(filepath.Base(exportPath), filepath.Ext(exportPath))
	ext := filepath.Ext(exportPath)
	return filepath.Join(dir, base+"_"+suffix+ext)
}

func containsInt(slice []int, v int) bool {
	for _, n := range slice {
		if n == v {
			return true
		}
	}
	return false
}
